File: app/src/main/java/com/dpis/module/settings/SettingsPresentationController.java
package com.dpis.module.settings;

import java.util.LinkedHashSet;
import java.util.Set;

/** JVM-testable state publisher over Java-owned Settings workflow execution. */
public final class SettingsPresentationController {
    public interface Port {
        SettingsUiState snapshot();
        void setSafeModeEnabled(boolean enabled);
        void setGlobalLogEnabled(boolean enabled);
        void setLauncherIconHidden(boolean hidden);
        void refresh();
    }
    public interface Listener { void onStateChanged(SettingsUiState state); }

    private final Port port;
    private final Set<Listener> listeners = new LinkedHashSet<>();

    public SettingsPresentationController(Port port) { this.port = port; }

    public void addListener(Listener listener) {
        if (listener == null) return;
        listeners.add(listener);
        listener.onStateChanged(port.snapshot());
    }

    public void removeListener(Listener listener) { listeners.remove(listener); }

    /**
     * Publishes only after the Java workflow reaches a stable observable state.
     * Confirmation dialogs and background work complete asynchronously, so the port
     * deliberately owns the timing rather than guessing from an action invocation.
     */
    public void publishState() { publish(port.snapshot()); }

    public void setSafeModeEnabled(boolean enabled) { port.setSafeModeEnabled(enabled); }
    public void setGlobalLogEnabled(boolean enabled) { port.setGlobalLogEnabled(enabled); }
    public void setLauncherIconHidden(boolean hidden) { port.setLauncherIconHidden(hidden); }
    public void refresh() { port.refresh(); }

    private void publish(SettingsUiState state) {
        for (Listener listener : new LinkedHashSet<>(listeners)) listener.onStateChanged(state);
    }
}

File: app/src/main/java/com/dpis/module/settings/SettingsPresentationController.kt
package com.dpis.module.settings

/** JVM-testable state publisher over Java-owned Settings workflow execution. */
class SettingsPresentationController(private val port: Port) {
    interface Port {
        fun snapshot(): SettingsUiState
        fun setSafeModeEnabled(enabled: Boolean)
        fun setGlobalLogEnabled(enabled: Boolean)
        fun setLauncherIconHidden(hidden: Boolean)
        fun refresh()
    }

    fun interface Listener {
        fun onStateChanged(state: SettingsUiState)
    }

    private val listeners = LinkedHashSet<Listener>()

    fun addListener(listener: Listener?) {
        if (listener == null) return
        listeners.add(listener)
        listener.onStateChanged(port.snapshot())
    }

    fun removeListener(listener: Listener?) {
        listeners.remove(listener)
    }

    /**
     * Publishes only after the Java workflow reaches a stable observable state.
     * Confirmation dialogs and background work complete asynchronously, so the port
     * deliberately owns the timing rather than guessing from an action invocation.
     */
    fun publishState() {
        publish(port.snapshot())
    }

    fun setSafeModeEnabled(enabled: Boolean) {
        port.setSafeModeEnabled(enabled)
    }

    fun setGlobalLogEnabled(enabled: Boolean) {
        port.setGlobalLogEnabled(enabled)
    }

    fun setLauncherIconHidden(hidden: Boolean) {
        port.setLauncherIconHidden(hidden)
    }

    fun refresh() {
        port.refresh()
    }

    private fun publish(state: SettingsUiState) {
        for (listener in LinkedHashSet(listeners)) listener.onStateChanged(state)
    }
}
